### PR TITLE
Coverage for 'gcloud._apitools.http_wrapper'

### PR DESCRIPTION
**Pull request by tseaver**
Uses #1193 as a base.

### CODE REVIEW

**Comment by tseaver** [comment]
@dhermes rebased after merge of #1193.

File: gcloud/_apitools/http_wrapper.py [hunk 1]
```diff
@@ -91,8 +90,7 @@ def _Httplib2Debuglevel(http_request, level, http=None):
     httplib2.debuglevel = old_level
     if http is not None:
         for connection_key, old_level in http_levels.items():
-            if connection_key in http.connections:
-                http.connections[connection_key].set_debuglevel(old_level)
+            http.connections[connection_key].set_debuglevel(old_level)
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: This changes the behavior of `apitools`

File: gcloud/_apitools/test_http_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+        self.assertEqual(_hits, [{'foo': 'bar'}])
+
+
+class _Dummy(object):
+    def __init__(self, **kw):
+        self.__dict__.update(kw)
+
+
+class _Request(object):
+    __slots__ = ('url', 'http_method', 'body', 'headers', 'loggable_body',)
```
Comment (dhermes) [inline gcloud/_apitools/test_http_wrapper.py]: Why necessary?

File: gcloud/_apitools/http_wrapper.py [hunk 3]
```diff
@@ -165,7 +163,7 @@ def ProcessContentRange(content_range):
             start, _, end = byte_range.partition('-')
             return int(end) - int(start) + 1
 
-        if '-content-encoding' in self.info and 'content-range' in self.info:
+        if 'content-encoding' in self.info and 'content-range' in self.info:
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: Why this change?

File: gcloud/_apitools/http_wrapper.py [hunk 4]
```diff
@@ -195,8 +193,7 @@ def CheckResponse(response):
     if response is None:
         # Caller shouldn't call us if the response is None, but handle anyway.
         raise exceptions.RequestError(
-            'Request to url %s did not return a response.' %
-            response.request_url)
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: Why ditch the more descriptive message?

File: gcloud/_apitools/http_wrapper.py [hunk 5]
```diff
@@ -223,10 +220,6 @@ def RebuildHttpConnections(http):
                 del http.connections[conn_key]
 
 
-def RethrowExceptionHandler(*unused_args):
-    raise
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: w00t! Delete, delete, delete!

File: gcloud/_apitools/http_wrapper.py [hunk 6]
```diff
@@ -245,14 +238,14 @@ def HandleExceptionsAndRebuildHttpConnections(retry_args):
                                    http_client.ResponseNotReady)):
         logging.debug('Caught HTTP error %s, retrying: %s',
                       type(retry_args.exc).__name__, retry_args.exc)
-    elif isinstance(retry_args.exc, socket.error):
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: Why ditch this? `socket.error` can be pretty generic.

File: gcloud/_apitools/test_http_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,824 @@
...
+
+    def test_w_connections(self):
+        connections = {'delete:me': object(), 'skip_me': object()}
+        http = _Dummy(connections=connections)
+        self._callFUT(http)
+        self.assertFalse('delete:me' in connections)
+        self.assertTrue('skip_me' in connections)
+
+
+class Test_HandleExceptionsAndRebuildHttpConnections(unittest2.TestCase):
```
Comment (dhermes) [inline gcloud/_apitools/test_http_wrapper.py]: Ew. What can be done here?

File: gcloud/_apitools/http_wrapper.py [hunk 8]
```diff
@@ -372,6 +321,53 @@ def _MakeRequestNoRetry(http, http_request, redirections=5,
     return response
 
 
+def MakeRequest(http, http_request, retries=7, max_retry_wait=60,
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: Why the move?

File: gcloud/_apitools/http_wrapper.py [hunk 9]
```diff
@@ -372,6 +321,53 @@ def _MakeRequestNoRetry(http, http_request, redirections=5,
...
+    """
+    retry = 0
+    while True:
+        try:
+            return wo_retry_func(
+                http, http_request, redirections=redirections,
+                check_response_func=check_response_func)
+        # retry_func will consume the exception types it handles and raise.
+        # pylint: disable=broad-except
+        except Exception as e:
```
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: This is scary. Worth dealing with?

File: gcloud/_apitools/http_wrapper.py (hunk 1 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: Not unless it is possible for `http.connections` to change during the `yield`:  `http_levels` is created by iterating its keys beforehand.

Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: It is certainly possible but I'm not familiar enough with the rest of the library to say what is safe here.

In most cases, you'd say "if the user removes the key before re-entry, they deserve the `KeyError`" but I don't know what sorts of operations users could perform or (more importantly) what those operations hide from the users.

File: gcloud/_apitools/test_http_wrapper.py (hunk 2 above)
Comment (tseaver) [inline gcloud/_apitools/test_http_wrapper.py]: I was trying to ensure that the code does not wander outside the "public" attrs of `Request`.

Comment (dhermes) [inline gcloud/_apitools/test_http_wrapper.py]: Seems overkill for a test, but I suppose it makes the test objects more efficient

File: gcloud/_apitools/http_wrapper.py (hunk 3 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: `'-content-encoding'` is not a valid HTTP header:  `apitools` has a typo here.

Re; `bytes` vs. `str`:  the response class follows `httplib2`.

File: gcloud/_apitools/http_wrapper.py (hunk 4 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: The check above it is for `response is None` -- we don't have the URL.

File: gcloud/_apitools/test_http_wrapper.py (hunk 7 above)
Comment (tseaver) [inline gcloud/_apitools/test_http_wrapper.py]: Fix it when doing PEP8 / lint cleanups.

File: gcloud/_apitools/http_wrapper.py (hunk 8 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: So that I can pass in the new `wo_retry_func` argument when testing.  That argument defaults to `_MakeRequestNoRetry`, which must therefore be defined before `MakeRequest`.

File: gcloud/_apitools/http_wrapper.py (hunk 9 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: During refactoring:  I'm planning to hoist the guts of `HandleExceptions...` to this point, so that each retryable / handleable exception type gets its own stanza (we won't need the bare `raise` then, either).

File: gcloud/_apitools/http_wrapper.py (hunk 4 above)
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: Haha! Testing FTW yet again. (I've been avoiding expanded the collapsed code.)

File: gcloud/_apitools/http_wrapper.py (hunk 3 above)
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: RE: "Type". Oy. I figured it was some suffix.

RE: Py2/Py3 annoyance. Should we comment to this effect?

File: gcloud/_apitools/test_http_wrapper.py (hunk 7 above)
Comment (dhermes) [inline gcloud/_apitools/test_http_wrapper.py]: OK

File: gcloud/_apitools/http_wrapper.py (hunk 8 above)
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: Ah nice. I figured it was due to the dependence but didn't realize the change to the method was material.

File: gcloud/_apitools/http_wrapper.py (hunk 9 above)
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: SGTM

File: gcloud/_apitools/http_wrapper.py (hunk 1 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: There are no users outside this module:  the (private) `_Httplib2DebugLevel` generator is a `contextlib.contextmanager` implementation, used only by `_MakeRequestNoRetry` to wrap around `http.request`, which pretty much has to be `httplib2.Http.request` (given the oddball signature).  That method *adds* keys to `self.connections`, but never removes them, which makes the check unnecessary.

File: gcloud/_apitools/test_http_wrapper.py (hunk 2 above)
Comment (tseaver) [inline gcloud/_apitools/test_http_wrapper.py]: I wouldn't be that defensive writing tests for new code, but `apitools` is more than a little whacky.

File: gcloud/_apitools/http_wrapper.py (hunk 3 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: > RE: Py2/Py3 annoyance. Should we comment to this effect?

I don't see such an issue inside this method -- I believe we will find some elsewhere when expanding the tests for coverage under Py3k.

File: gcloud/_apitools/http_wrapper.py (hunk 6 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: Yup, the others were unreachable until I moved it down.

File: gcloud/_apitools/http_wrapper.py (hunk 5 above)
Comment (tseaver) [inline gcloud/_apitools/http_wrapper.py]: Yup.  Once we get everything covered, and can identify just the paths we need in `gcloud.storage` / `gcloud.bigquery`, there will be a bunch more deletion to be done.

File: gcloud/_apitools/http_wrapper.py (hunk 1 above)
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: OK

File: gcloud/_apitools/test_http_wrapper.py (hunk 2 above)
Comment (dhermes) [inline gcloud/_apitools/test_http_wrapper.py]: Sure thing

File: gcloud/_apitools/http_wrapper.py (hunk 3 above)
Comment (dhermes) [inline gcloud/_apitools/http_wrapper.py]: OK

**Comment by dhermes** [comment]
LGTM
